packages/@uppy/aws-s3-multipart/src/MultipartUploader.js

```diff
@@ -29,429 +25,121 @@ function ensureInt (value) {
-        this.partsInProgress -= 1
-      },
-    })
+  #resumeUpload () {
+    this.#uploadPromise = this
+      .options.companionComm.resumeUploadFile(this.#file, this.#chunks, this.#abortController.signal)
+      .then(this.#onSuccess, this.#onReject)
   }
 
-  #uploadPart (index, prePreparedPart) {
-    const valid = typeof prePreparedPart?.url === 'string'
-    if (!valid) {
-      throw new TypeError('AwsS3/Multipart: Got incorrect result for `prePreparedPart`, expected an object `{ url }`.')
-    }
-
-    const { url, headers } = prePreparedPart
-    if (this.#aborted()) {
-      throw createAbortError()
-    }
+  #onPartProgress = (index) => (ev) => {
+    if (!ev.lengthComputable) return
 
-    return this.#uploadPartBytes(index, url, headers)
-  }
-
-  #onPartProgress (index, sent) {
-    this.chunkState[index].uploaded = ensureInt(sent)
+    const sent = ev.loaded
+    this.#chunkState[index].uploaded = ensureInt(sent)
 
-    const totalUploaded = this.chunkState.reduce((n, c) => n + c.uploaded, 0)
-    this.options.onProgress(totalUploaded, this.file.size)
+    const totalUploaded = this.#chunkState.reduce((n, c) => n + c.uploaded, 0)
+    this.options.onProgress(totalUploaded, this.#file.size)
   }
 
-  #onPartComplete (index, etag) {
-    this.chunkState[index].etag = etag
-    this.chunkState[index].done = true
+  #onPartComplete = (index) => (etag) => {
+    this.#chunkState[index].etag = etag
+    this.#chunkState[index].done = true
 
     const part = {
       PartNumber: index + 1,
       ETag: etag,
     }
-    this.parts.push(part)
-
     this.options.onPartComplete(part)
   }
 
-  #uploadPartBytes (index, url, headers) {
-    const body = this.chunks[index]
-    const { signal } = this.abortController
-
-    let defer
-    const promise = new Promise((resolve, reject) => {
-      defer = { resolve, reject }
-    })
-
-    const xhr = new XMLHttpRequest()
-    xhr.open('PUT', url, true)
-    if (headers) {
-      Object.keys(headers).forEach((key) => {
-        xhr.setRequestHeader(key, headers[key])
-      })
-    }
-    xhr.responseType = 'text'
-
-    function cleanup () {
-      // eslint-disable-next-line no-use-before-define
-      signal.removeEventListener('abort', onabort)
-    }
-    function onabort () {
-      xhr.abort()
-    }
-    signal.addEventListener('abort', onabort)
-
-    xhr.upload.addEventListener('progress', (ev) => {
-      if (!ev.lengthComputable) return
-
-      this.#onPartProgress(index, ev.loaded, ev.total)
-    })
-
-    xhr.addEventListener('abort', () => {
-      cleanup()
-
-      defer.reject(createAbortError())
-    })
-
-    xhr.addEventListener('load', (ev) => {
-      cleanup()
-
-      if (ev.target.status < 200 || ev.target.status >= 300) {
-        const error = new Error('Non 2xx')
-        error.source = ev.target
-        defer.reject(error)
-        return
-      }
-
-      // This avoids the net::ERR_OUT_OF_MEMORY in Chromium Browsers.
-      this.chunks[index] = null
```

Comment:
Are we still doing this correctly somewhere else

Reply:
Good catch, f00f98627cbb6697b4d95784516b97f515607a9a